### PR TITLE
Add test for unmarshalling dataclasses into codegen structs

### PR DESCRIPTION
**Pull request by rooftoofwoof**
This PR primarily adds a test for unmarshalling dataclasses into codegen structs. With some supporting refactors:
- [x] Extract recursive logic for setting Arrow builders
- [x] Add logic for StructBuilder
- [x] Handle building an Arrow table for a slice of pointers (a dataclass with 4 pointer fields would have 4 underlying slices of pointers)

### CODE REVIEW

File: internal/feather.go [hunk 1]
```diff
@@ -57,21 +58,170 @@ func convertReflectToArrowType(value reflect.Type) (arrow.DataType, error) {
...
+	}
+	if slice.Len() == 0 {
+		return nil // No values to append
+	}
+
+	elemType := valType.Elem()
+	elemKind := elemType.Kind()
+	values := slice.Interface()
+	switch elemKind {
+	case reflect.Ptr:
```
Comment (rooftoofwoof) [inline internal/feather.go:112]: This is new

File: internal/feather.go [hunk 2]
```diff
@@ -57,21 +58,170 @@ func convertReflectToArrowType(value reflect.Type) (arrow.DataType, error) {
...
+		builder.(*array.BooleanBuilder).AppendValues(values.([]bool), nullMask)
+	case reflect.Struct:
+		if elemType == reflect.TypeOf(time.Time{}) {
+			timeSlice := values.([]time.Time)
+			timestampSlice := make([]arrow.Timestamp, 0, len(timeSlice))
+			for _, t := range timeSlice {
+				timestampSlice = append(timestampSlice, arrow.Timestamp(t.UnixNano()))
+			}
+			builder.(*array.TimestampBuilder).AppendValues(timestampSlice, nullMask)
+		} else {
```
Comment (rooftoofwoof) [inline internal/feather.go:173]: This else case is new

File: internal/feather.go [hunk 3]
```diff
@@ -57,21 +58,170 @@ func convertReflectToArrowType(value reflect.Type) (arrow.DataType, error) {
...
+				)
+			}
+
+			var columns []reflect.Value
+			for j := 0; j < numFieldsReflect; j++ {
+				structz := slice.Index(0)
+				sliceType := reflect.SliceOf(structz.Field(j).Type())
+				col := reflect.MakeSlice(sliceType, 0, slice.Len())
+				for i := 0; i < slice.Len(); i++ {
+					col = reflect.Append(col, structz.Field(j))
```
Comment (rooftoofwoof) [inline internal/feather.go:204]: this seems funky

File: internal/feather.go [hunk 4]
```diff
@@ -57,21 +58,170 @@ func convertReflectToArrowType(value reflect.Type) (arrow.DataType, error) {
...
+				}
+				columns = append(columns, col)
+			}
+
+			for i := 0; i < structBuilder.NumField(); i++ {
+				fieldBuilder := structBuilder.FieldBuilder(i)
```
Comment (rooftoofwoof) [inline internal/feather.go:210]: inline?

File: internal/feather.go [hunk 5]
```diff
@@ -57,21 +58,161 @@ func convertReflectToArrowType(value reflect.Type) (arrow.DataType, error) {
...
+				timestampSlice = append(timestampSlice, arrow.Timestamp(t.UnixNano()))
+			}
+			builder.(*array.TimestampBuilder).AppendValues(timestampSlice, nullMask)
+		} else {
+			sBuilder, ok := builder.(*array.StructBuilder)
+			if !ok {
+				return errors.Errorf("internal error: expected struct builder, found %T", builder)
+			}
+
+			numFieldsReflect := slice.Index(0).NumField()
```
Comment (Nathan-Fenner) [inline internal/feather.go:179]: Replace this with

```go
numFieldsReflect := valType.NumField()
```

Comment (rooftoofwoof) [reply internal/feather.go:179]: With that I'm getting 
```
reflect: NumField of non-struct type []chalk.testLatLng
```

Comment (rooftoofwoof) [reply internal/feather.go:179]: ah `valType.Elem().NumField()` did it

Comment (rooftoofwoof) [reply internal/feather.go:179]: thank you!

File: internal/feather.go [hunk 6]
```diff
@@ -57,21 +58,161 @@ func convertReflectToArrowType(value reflect.Type) (arrow.DataType, error) {
...
+						"found %d fields in struct and %d fields in Arrow struct schema",
+					numFieldsReflect,
+					numFieldsArrow,
+				)
+			}
+
+			var columns []reflect.Value
+			for j := 0; j < numFieldsReflect; j++ {
+				firstStruct := slice.Index(0)
+				sliceType := reflect.SliceOf(firstStruct.Field(j).Type())
```
Comment (Nathan-Fenner) [inline internal/feather.go:193]: Replace this with

```go
sliceTypeForField := reflect.SliceOf(valType.Field(j).Type)
```

File: internal/feather.go [hunk 7]
```diff
@@ -57,21 +58,161 @@ func convertReflectToArrowType(value reflect.Type) (arrow.DataType, error) {
...
+				col := reflect.MakeSlice(sliceType, 0, slice.Len())
+				for i := 0; i < slice.Len(); i++ {
+					col = reflect.Append(col, slice.Index(i).Field(j))
+				}
+				columns = append(columns, col)
+			}
+
+			for i := 0; i < sBuilder.NumField(); i++ {
```
Comment (Nathan-Fenner) [inline internal/feather.go:201]: You are missing validation to verify that the field names/order for `sBuilder` are the same as the struct type, I think

File: internal/feather.go [hunk 8]
```diff
@@ -57,21 +58,161 @@ func convertReflectToArrowType(value reflect.Type) (arrow.DataType, error) {
...
+				)
+			}
+
+			var columns []reflect.Value
+			for j := 0; j < numFieldsReflect; j++ {
+				firstStruct := slice.Index(0)
+				sliceType := reflect.SliceOf(firstStruct.Field(j).Type())
+				col := reflect.MakeSlice(sliceType, 0, slice.Len())
+				for i := 0; i < slice.Len(); i++ {
+					col = reflect.Append(col, slice.Index(i).Field(j))
```
Comment (Nathan-Fenner) [inline internal/feather.go]: This will copy the values, so when this new slice is modified, the original slice will not be affected, so the values will get lost, I think? The test should catch this, but it looks like it's not currently asserting equality.

File: internal/feather.go (hunk 7 above)
Comment (Nathan-Fenner) [reply internal/feather.go:201]: Ideally, you'd also allow for renaming via some kind of struct tag, like

```go
type MyCustomType struct {
  ID       `arrow_column:"id"`
  IsActive `arrow_column:"is_active"`
}

```

File: internal/feather.go (hunk 6 above)
Comment (rooftoofwoof) [reply internal/feather.go]: addressed - thanks!

File: internal/feather.go (hunk 7 above)
Comment (rooftoofwoof) [reply internal/feather.go:222]: addressed

**Review by Nathan-Fenner — APPROVED** [review]

**Comment by Nathan-Fenner** [comment]
Please add a test for the null logic, so that we make sure that it stays right. Maybe also a test that combines optional struct, and struct with optional fields.

File: internal/utils.go [hunk 9]
```diff
@@ -88,3 +91,42 @@ func getFeatureNameFromFqn(fqn string) string {
 	lastPart := strings.Split(fqn, ".")
 	return lastPart[len(lastPart)-1]
 }
+
+// ChalkpySnakeCase aims to be in parity with
+//
+//	our Python implementation of snake_case
```
Comment (rooftoofwoof) [inline internal/utils.go:97]: what is this
